Sources/Core/Sources/ECS/Singles/InputState.swift

```diff
@@ -86,11 +94,29 @@ public final class InputState: SingleComponent {
   /// Ticks the input state by flushing ``newlyPressed`` into ``keys`` and ``inputs``, and clearing
   /// ``newlyReleased``. Also emits events to the given ``EventBus``.
   func tick(_ isInputSuppressed: [Bool], _ eventBus: EventBus) {
+    /// increment the tick count
+    tickCount += 1
+    /// increment the time since the forwards key was pressed if it is currently pressed
     assert(isInputSuppressed.count == newlyPressed.count, "`isInputSuppressed` should be the same length as `newlyPressed`")
     for (var event, suppressInput) in zip(newlyPressed, isInputSuppressed) {
       if suppressInput {
         event.input = nil
       }
+      ///test for forwards key
+      if event.input == .moveForward {
+        /// if the forwards key has been pressed, released, and pressed again within 6 ticks, sprint
+        if forwardsDownTime < forwardsUpTime && (forwardsDownTime + 6) > tickCount {
+          inputs.insert(.sprint)
+        } else {
+          /// if not then stop the user from sprinting
+          inputs.remove(.sprint)
+        }
+        /// reset the counts so the ints don't overflow
+        tickCount = 0
+        forwardsDownTime = 0
```

Comment:
I think `forwardsDownTime` isn't necessary, you just need to store the tick at which forwards was last released, and compare it to the current tick count whenever forwards is pressed.

Reply:
it's necessary to track whether the user holds down the forwards key. it's counted as multiple keypresses and will sprint if they just hold it down. the forwardsDownTime is compared to forwardsUpTime to make sure they have released the key.